bamboo/unit_tests/test_unit_datastore_imagenet.py

```diff
@@ -114,8 +116,97 @@ def construct_data_reader(lbann):
+    '''Executes the input test function  
 
-def create_test_func(baseline_test_func, datastore_test_func):
+    Args:
+        run_datastore_test_func (function): test function 
+        baseline_metrics: list of metrics against which the output of
+                          the test function will be compared
+        profile_data: dictionary of key, value pairs for testing
+                      entries in the output file: data_store_profile_train.txt
+
+    Returns:
+        list containg test name, pass/fail, etc.
+        On error, this will have the form:
+          ['FAILED', <test function name>, <error>]
+        on success:
+          ['passed', <test function name>]
+    '''
+    datastore_test_output = test_func(cluster, exes, dirname)
+
+    test_name = test_func.__name__
+    r = ['passed', test_name]
+    datastore_metrics = []
+    with open(datastore_test_output['stdout_log_file']) as f:
+        for line in f:
+            match = re.search('validation metric : ([0-9.]+)', line)
+            if match:
+                datastore_metrics.append(float(match.group(1)))
+
+    # Check if metrics are same in baseline and data store experiments
+    # Note: "Print statistics" callback will print up to 6 digits
+    # of metric values.
+    if len(baseline_metrics) != len(datastore_metrics) :
+        r[0] = 'FAILED'
+        r.append('baseline and data store experiments did not run for same number of epochs; num baseline: ' + str(len(baseline_metrics)) + '; num ds: ' + str(len(datastore_metrics)))
+
+    for i in range(len(datastore_metrics)):
+        x = baseline_metrics[i]
+        xhat = datastore_metrics[i]
+        eps = np.finfo(np.float32).eps
+        ceillogx = int(math.ceil(math.log10(x)))
+        if abs(x-xhat) >= max(8*eps*x, 1.5*10**(ceillogx-6)) :
+            r[0] = 'FAILED'
+            r.append('found large discrepancy in metrics for baseline and data store experiments')
+
+    # Check if entries profile_data exist and have correct values
+    d = None
+    for key in profile_data.keys() :
+      if test_name.find(key) != -1 :
+        d = profile_data[key]
+        break
+    assert d != None, 'failed to find key for profile_data'    
+
+    found_profile_data = {}
+    with open(datastore_test_output['work_dir'] + '/data_store_profile_train.txt') as f:
+        for line in f:
+            for key in d :
+                if key in line and key not in found_profile_data.keys() :
+                    t = line.split()
+                    found_profile_data[key] = t[-1]
+
+    for key in d.keys() :
+        if key not in found_profile_data.keys() :
+            r[0] = 'FAILED'
+            r.append('missing key in profile_data: ' + key)
+        elif found_profile_data[key] != d[key] :
+            r[0] = 'FAILED'
+            r.append('bad value for "' + key + '; value is: ' + str(found_profile_data[key]) + '; should be: ' + str(d[key]))
+    return r
```

Comment:
It would be simpler to just have `assert`s inside this function rather than returning a list of error messages. See my earlier code suggestions.